gptel-context.el

```diff
@@ -195,38 +211,85 @@ Return PATH if added, nil if ignored."
     (message "Ignoring unsupported binary file \"%s\"." path)
     nil))
 
-(defun gptel-context--add-directory (path action)
+(defun gptel-context--add-directory (path action &optional project-files)
   "Process all files in directory at PATH according to ACTION.
-ACTION should be either `add' or `remove'."
-  (let ((files (directory-files-recursively path "." t)))
-    (mapc (lambda (file)
-            (unless (file-directory-p file)
-              (pcase-exhaustive action
-                ('add
-                 (if (gptel--file-binary-p file)
-                     (gptel-context--add-binary-file file)
-                   (gptel-context--add-text-file file)))
-                ('remove
-                 (setf (alist-get file gptel-context--alist nil 'remove #'equal) nil)))))
-          files)
+ACTION should be either `add' or `remove'.
+
+PROJECT-FILES is a list of all files in the current project."
+  (let* ((project-files (or project-files
+			    (and (eq action 'add)
+				 (gptel-context--get-project-files path))))
+	 (root (car-safe project-files)))
+    (dolist (file (directory-files-recursively path "." t))
+      (unless (file-directory-p file)
+	(if (and gptel-context-restrict-to-project-files
+		 root
+		 (eq action 'add)
+		 (not (member file project-files)))
+	    (gptel-context--message-skipped file project-files)
+	  (pcase-exhaustive action
+	    ('add
+	     (if (gptel--file-binary-p file)
+		 (gptel-context--add-binary-file file)
+	       (gptel-context--add-text-file file)))
+	    ('remove
+	     (setf (alist-get file gptel-context--alist nil 'remove #'equal) nil))))))
     (when (eq action 'remove)
       (message "Directory \"%s\" removed from context." path))))
 
-(defun gptel-context-add-file (path)
+(defun gptel-context-add-file (path &optional project-files)
```

Comment:
It seems pretty odd (to me) to expose this in public API. I would expect this function (just going off its name) to add the named file.

If this is intended to opt into / out of automatically excluding ignored files, I would kinda expect that to be controlled by let-binding the option. This differs from codebase to codebase, though, so I might be off there.